File: runtime/CSlib/output_stream_port.cs
using System;
using System.IO;

namespace bigloo {
   public class output_stream_port: output_port {
      protected Stream stream;

      public output_stream_port( Stream  stream, byte[]  name ) : base( name ) {
	 this.stream= stream;
      }

      public output_stream_port( Stream  stream ) {
	 this.stream= stream;
      }

      public output_stream_port( byte[] file ) : base( file ) {
	 string file_name= foreign.newstring( file );

	 try {
	    // should we keep this try/catch block even
	    // when Mono bug is be corrected ?
	    if( file_name.ToLower().Equals( "null:" ) )
	       stream= Stream.Null;
	    else
	       stream= new FileStream( file_name,
				       FileMode.OpenOrCreate,
				       FileAccess.Write,
				       FileShare.Read );
	 } catch ( Exception ) {
	    foreign.fail( "open_output_file",
			  "Could not open file " + file_name,
			  file );
	 }
      }

      public output_stream_port( byte[] file, bool append ) : base( file ) {
	 try {
	    // !!!!! is this try/catch block to be kept
	    // or just a protecting against a Mono/PNet bug ?????
	    stream= File.Open( foreign.newstring( file ),
			       (append ? FileMode.Append : FileMode.Create),
			       FileAccess.Write );
	 } catch ( Exception e ) {
	    foreign.fail( "append output port", e.Message, file );
	 }
      }

      public override Object close() {
	 try {
	    stream.Close();
	    base.close();
	 } catch ( Exception e ) {
	    if( stream == null )
	       return bigloo.foreign.BUNSPEC;
	    else
	       foreign.fail( "close", e.Message, this );
	 }

	 return this;
      }

      public override void invoke_flush_hook( bigloo.bint size ) {
	 if( fhook is procedure ) {
	    Object s = ((procedure)fhook).funcall2( this, size );

	    if( s is byte[] ) {
	       stream.Write( (byte [])s, 0, ((byte [])s).Length );
	    } else {
	       if( s is bigloo.bint &&
		   flushbuf is byte[] &&
		   bigloo.foreign.CINT( (bigloo.bint)s ) <= ((byte[])flushbuf).Length ) {
		  stream.Write( (byte[])flushbuf,
				0,
				bigloo.foreign.CINT( (bigloo.bint)s ) );
	       }
	    }
	 }
      }
      
      public override Object flush() {
	 try {
	    invoke_flush_hook( bigloo.foreign.BINT( 0 ) );
	    stream.Flush();
	    return bbool.vrai;
	 } catch ( Exception e ) {
	    if( stream != null ) foreign.fail( "flush", e.Message, this );
	    return bbool.faux;
	 }
      }

      static byte[] mono_WriteByte_workaround = new byte[ 1 ];
    
      public override void write( int cn ) {
	 try {
	    invoke_flush_hook( bigloo.foreign.BINT( 1 ) );
	    mono_WriteByte_workaround[ 0 ] = (byte)cn;
	    stream.Write( mono_WriteByte_workaround, 0, 1 );
	    stream.Flush();
	 } catch ( Exception e ) {
	    if( stream != null ) foreign.fail( "write", e.Message, this );
	 }
      }

      public override void write( byte[] s ) {
	 try {
	    // CARE pnet bug
	    int n = s.Length;

	    invoke_flush_hook( bigloo.foreign.BINT( n ) );
	
	    if( n == 3294 )
	       for ( int i= 0 ; i < 3294 ; ++i )
		  write( s[ i ] );
	    else
	       stream.Write( s, 0, s.Length );
	    stream.Flush();
	 } catch ( Exception e ) {
	    if( stream != null ) foreign.fail( "write", e.Message, this );
	 }
      }

      public override void write( byte[] s, int start, int end ) {
	 try {
	    invoke_flush_hook( bigloo.foreign.BINT( end - start ) );
	    stream.Write( s, start, end - start );
	    stream.Flush();
	 } catch( Exception e ) {
	    if( stream != null ) foreign.fail( "write", e.Message, this );
	 }
      }

      public override void write( String s ) {
	 try {
	    int len = s.Length;

	    invoke_flush_hook( bigloo.foreign.BINT( len ) );
	
	    for( int i= 0 ; i < len ; ++i ) {
	       mono_WriteByte_workaround[ 0 ] = (byte)s[i];
	       stream.Write( mono_WriteByte_workaround, 0, 1 );
	    }
	 } catch( Exception e ) {
	    if( stream != null ) foreign.fail( "write", e.Message, this );
	 }
      }
   }
}

File: runtime/Jlib/output_stream_port.java
/*=====================================================================*/
/*    .../project/bigloo/5.0a/runtime/Jlib/output_stream_port.java     */
/*    -------------------------------------------------------------    */
/*    Author      :  manuel serrano                                    */
/*    Creation    :  Tue Apr 28 18:22:05 2026                          */
/*    Last change :  Thu Apr 30 12:40:18 2026 (serrano)                */
/*    Copyright   :  2026 manuel serrano                               */
/*    -------------------------------------------------------------    */
/*    Output stream ports                                              */
/*=====================================================================*/
package bigloo;

import java.io.*;

/*---------------------------------------------------------------------*/
/*    output_stream_port ...                                           */
/*---------------------------------------------------------------------*/
public class output_stream_port extends output_port {
   OutputStream out;

   public output_stream_port(byte[] name) {
      super(name);
   }

   public output_stream_port(OutputStream _out) {
      super();
      out = _out;
   }
   
   public output_stream_port(OutputStream _out, byte[] name) {
      super(name);
      out = _out;
   }
   
   public Object close() {
      try {
         out.close();
      } catch(Throwable _t) {
         ;
      }

      return super.close();
   }
   
   public Object flush() {
      try {
         out.flush();
         return bbool.vrai;
      } catch (final Exception e) {
         if (out != null)
            foreign.fail("flush", e, this);
         return bbool.faux;
      }
   }

   public Object bgl_output_port_seek(final int pos) throws IOException {
      return bigloo.foreign.BFALSE;
   }

   protected void invoke_flush_hook(bigloo.bint size) {
      if (fhook instanceof procedure) {
	 Object s = ((procedure)fhook).funcall2(this, size);

	 try {
	    if (s instanceof byte[]) {
	       out.write((byte [])s, 0, ((byte [])s).length);
	    } else {
	       if (s instanceof bigloo.bint &&
		   flushbuf instanceof byte[] &&
		   bigloo.foreign.CINT((bigloo.bint)s) <= ((byte[])flushbuf).length &&
		   bigloo.foreign.CINT((bigloo.bint)s) > 0)
	       {
		  out.write((byte[])flushbuf,
			    0,
			    bigloo.foreign.CINT((bigloo.bint)s));
	       }
	       
	    }
	 } catch(Throwable _t) {
	    ;
	 }
      }
   }
      
   public void write(final int cn) {
      try {
	 invoke_flush_hook(bigloo.foreign.BINT(1));
	 out.write(cn);
      } catch (final Exception e) {
	 if (out != null)
	    foreign.fail("write", e, this);
      }
   }

   public void write(final byte[] s) {
      try {
	 invoke_flush_hook(bigloo.foreign.BINT(s.length));
	 out.write(s, 0, s.length);
      } catch (final Exception e) {
	 if (out != null)
	    foreign.fail("write", e, this);
      }
   }

   public void write(final byte[] s, int offset, int len) {
      try {
	 invoke_flush_hook(bigloo.foreign.BINT(len - offset));
	 out.write(s, offset, (len - offset));
      } catch (final Exception e) {
	 if (out != null)
	    foreign.fail("write", e, this);
      }
   }

   public void write(final String s) {
      try {
	 final int len = s.length();

	 invoke_flush_hook(bigloo.foreign.BINT(len));
	 
	 for (int i= 0 ;i < len ; ++i)
	    out.write((byte)s.charAt(i));
      } catch (final Exception e) {
	 if (out != null)
	    foreign.fail("write", e, this);
      }
   }

   public void write(final output_port p) {
      p.write("#<output_port:" + new String(name) + ">");
   }

   public boolean truncate(long size) {
      if (out instanceof FileOutputStream) {
	 try {
	    return JDK.truncate((FileOutputStream)out, size);
	 } catch(Exception _e) {
	    return false;
	 }
      } else {
	 return false;
      }
   }

   public OutputStream getOutputStream() {
      return out;
   }
}
